plugins/modem-manager/fu-qmi-pdc-updater.c

```diff
@@ -0,0 +1,432 @@
+
+static void
+fu_qmi_pdc_updater_qmi_device_open_abort_ready (GObject *qmi_device, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
+	g_assert (ctx->error != NULL);
+
+	/* ignore errors when aborting open */
+	qmi_device_close_finish (QMI_DEVICE (qmi_device), res, NULL);
+
+	ctx->open_attempts--;
+	if (ctx->open_attempts == 0) {
+		g_clear_object (&ctx->qmi_client);
+		g_clear_object (&ctx->qmi_device);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	/* retry */
+	g_clear_error (&ctx->error);
+	fu_qmi_pdc_updater_qmi_device_open_attempt (ctx);
+}
+
+static void
+fu_qmi_pdc_updater_open_abort (OpenContext *ctx)
+{
+	qmi_device_close_async (ctx->qmi_device,
+				15, NULL, fu_qmi_pdc_updater_qmi_device_open_abort_ready, ctx);
+}
+
+static void
+fu_qmi_pdc_updater_qmi_device_allocate_client_ready (GObject *qmi_device, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
+	ctx->qmi_client = QMI_CLIENT_PDC (qmi_device_allocate_client_finish (QMI_DEVICE (qmi_device), res, &ctx->error));
+	if (ctx->qmi_client == NULL) {
+		fu_qmi_pdc_updater_open_abort (ctx);
+		return;
+	}
+
+	g_main_loop_quit (ctx->mainloop);
+}
+
+static void
+fu_qmi_pdc_updater_qmi_device_open_ready (GObject *qmi_device, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
+	if (!qmi_device_open_finish (QMI_DEVICE (qmi_device), res, &ctx->error)) {
+		fu_qmi_pdc_updater_open_abort (ctx);
+		return;
+	}
+
+	qmi_device_allocate_client (ctx->qmi_device, QMI_SERVICE_PDC, QMI_CID_NONE, 5, NULL,
+				    fu_qmi_pdc_updater_qmi_device_allocate_client_ready, ctx);
+}
+
+static void
+fu_qmi_pdc_updater_qmi_device_open_attempt (OpenContext *ctx)
+{
+	QmiDeviceOpenFlags open_flags = QMI_DEVICE_OPEN_FLAGS_NONE;
+
+	g_debug ("trying to open QMI device...");
+
+	/* automatically detect QMI and MBIM ports */
+	open_flags |= QMI_DEVICE_OPEN_FLAGS_AUTO;
+	/* qmi pdc requires indications, so enable them by default */
+	open_flags |= QMI_DEVICE_OPEN_FLAGS_EXPECT_INDICATIONS;
+	/* all communication through the proxy */
+	open_flags |= QMI_DEVICE_OPEN_FLAGS_PROXY;
+
+	qmi_device_open (ctx->qmi_device, open_flags, 5, NULL,
+			 fu_qmi_pdc_updater_qmi_device_open_ready, ctx);
+}
+
+static void
+fu_qmi_pdc_updater_qmi_device_new_ready (GObject *source, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
+	ctx->qmi_device = qmi_device_new_finish (res, &ctx->error);
+	if (ctx->qmi_device == NULL) {
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	fu_qmi_pdc_updater_qmi_device_open_attempt (ctx);
+}
+
+gboolean
+fu_qmi_pdc_updater_open (FuQmiPdcUpdater *self, GError **error)
+{
+	g_autoptr(GMainLoop) mainloop = g_main_loop_new (NULL, FALSE);
+	g_autoptr(GFile) qmi_device_file = g_file_new_for_path (self->qmi_port);
+	OpenContext ctx = {
+		.mainloop = mainloop,
+		.open_attempts = FU_QMI_PDC_MAX_OPEN_ATTEMPTS,
+	};
```

Comment:
don't we need some other fields initialized here?

Reply:
ok